content/blog/migrating-a-cloud-application-to-kubernetes/index.md

```diff
@@ -0,0 +1,476 @@
+
+```typescript
+const databaseAppName = "database-side-service";
+const databaseAppLabels = { appClass: databaseAppName };
+const databaseDeployment = new k8s.apps.v1.Deployment(databaseAppName, {
+    metadata: { name: databaseAppName, labels: databaseAppLabels },
+    spec: {
+        replicas: 1,
+        selector: { matchLabels: databaseAppLabels },
+        template: {
+            metadata: { labels: databaseAppLabels },
+            spec: {
+                containers: [{
+                    name: databaseAppName,
+                    image: awsx.ecr.buildAndPushImage("database-side-service", "./databaseside").image(),
+                    ports: [{ name: "http", containerPort: 5432 }],
+                    env: [
+                        { name: "ADMIN_NAME", value: sqlAdminName },
+                        { name: "ADMIN_PASSWORD", value: sqlAdminPassword },
+                    ],
+                    volumeMounts: [{
+                        name: "persistent-volume",
+                        mountPath: "/persistentVolume"
+                    }],
+                    resources: {
+                        limits: {
+                            memory: "1Gi",
+                            cpu: "1000m"
+                        }
+                    }
+                }],
+                volumes: [{
+                    name: "persistent-volume",
+                    awsElasticBlockStore: {
+                        volumeID: ebsVolume.id,
+                    },
+                }],
+                affinity: {  
+                    nodeAffinity: {
+                        requiredDuringSchedulingIgnoredDuringExecution: {
+                            nodeSelectorTerms: [{
+                                matchExpressions: [{
+                                    key: "failure-domain.beta.kubernetes.io/zone",
+                                    operator: "In",
+                                    values: [ebsVolume.availabilityZone],
+                                }]
+                            }]
+                        }
+                    }
+                }
+            }
+        },
+        strategy: {
+            type: "Recreate"
+        }
+    }}, {
+    deleteBeforeReplace: true,
+    provider: eksCluster.provider,
+});
+```
+
+In order to have a clean, simple URL through which we can access the database, we create a Kubernetes service for it.
+
+```typescript
+const databasesideListener = new k8s.core.v1.Service("database-side-listener", {
+    metadata: { labels: databaseAppLabels },
+    spec: {
+        type: "LoadBalancer",
+        ports: [{ port: 5432, targetPort: "http" }],
+        selector: databaseAppLabels,
+    }}, {
+    provider: eksCluster.provider,
+    }
+);
+
```

Comment:
This needs to be changed in the project. A [load balancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) would expose PGSQL outside of the cluster, which is bad practice. Connection to the database should only use clusterIPs, if the application is in the same Pod/Node you can use selectors.

Reply:
It's definitely true that exposing PGSQL to the internet is bad practice, but the Pulumi PostgreSQL provider that I used seems to require a direct server address to connect to. If no external address is given to the database, then I don't think the provider is able to function.

Reply:
Removed the providers entirely, and instead changed the project to perform the initialization steps in the container bash script. The database now uses a ClusterIP service, and is only accessible from within the cluster.